Restructuring

ultronozm opened the pull request.

skissue commented on llm-tool-collection.el:
```diff
@@ -35,89 +35,133 @@
+This macro creates a constant with the tool's specs and defines a
+function under `llm-tc/NAME' whose docstring is the value of the spec
+`:description'.  The tool is then added to
+`llm-tool-collection--all-tools'."
```
If this is in the docstring, should we make it into a public variable (as in, `llm-tool-collection-all-tools`)? I originally intended for this to be an internal variable only (for `llm-tool-collection-get-all`), but I could see how a list of the symbols could be useful too.

ultronozm replied:
I decided to leave it private for now, tweaking the docstring

skissue commented on llm-tool-collection.el:
```diff
@@ -35,89 +35,133 @@
+    (string-replace "-" "_" (symbol-name name))))
 
-(defmacro llm-tool-collection-deftool (name specs arguments &rest body)
+(defvar llm-tool-collection-after-tool-define-hook nil
```
The [Emacs Lisp manual recommends](https://www.gnu.org/software/emacs/manual/html_node/emacs/Hooks.html) that we use `-functions` instead of `-hook` for "abnormal" hooks (those that take arguments).

ultronozm replied:
Thanks, fixed

skissue commented on llm-tool-collection.el:
```diff
@@ -35,89 +35,133 @@
+           ,(concat (plist-get specs :description) "\n\n"
+                    "Definition generated by `llm-tool-collection'.")
+           ,@body)
+         (add-to-list 'llm-tool-collection--all-tools ',sym)
```
The docstring for `add-to-list` states:

> This is meant to be used for adding elements to configuration
variables, such as adding a directory to a path variable
like load-path, but please do not abuse it to construct
arbitrary lists in Elisp code, where using push or cl-pushnew
will get you more efficient code.

Thus, would it be better to use `cl-pushnew` instead? I know you explicitly wanted to avoid the dependency on `cl-lib`, but I don't think it's a big deal, given that it would be a compile-time dependency. However, I'm willing to defer to popular convention here—if `add-to-list` is commonly used in packages, then we can follow others' example. I suppose this is also a minor, small use, so perhaps I'm overreacting :).

ultronozm replied:
Thanks, fixed (by requiring cl-lib)

skissue reviewed (commented):
Thanks so much! One last nitpick—would you be willing to edit the commit messages to adhere to the format of the previous commits (based on [Conventional Commits](https://www.conventionalcommits.org/))?

ultronozm commented:
These all sound good and I’ll implement them later, although for the commit
messages, I’d have a slight preference for adopting the GNU ChangeLog
style, mainly out of laziness - that style is what Emacs and nearby
packages (gptel, llm, …) use, and the built-in support for drafting such
messages is good (log-edit-generate-changelog-from-diff, etc), while I
wouldn’t know how to be comparably lazy with the conventions you mention -
do you?

On Sat, 22 Mar 2025 at 14:41, Ad ***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> Thanks so much! One other nitpick—would you be willing to edit the commit
> messages to adhere to the format of the previous commits (based on Conventional
> Commits <https://www.conventionalcommits.org/>)?
> ------------------------------
>
> In llm-tool-collection.el
> <https://github.com/skissue/llm-tool-collection/pull/2#discussion_r2008772967>
> :
>
> > -         ***@***.***)
> -       (push ',sym llm-tool-collection--all-tools))))
> -
> +ARGS is a list where each element is of the form
> +
> +  (ARGNAME :type TYPE :description \"DESCRIPTION\").
> +
> +Arguments after the special symbol `&optional' are marked with
> +`:optional t`.
> +
> +BODY contains the function body.
> +
> +This macro creates a constant with the tool's specs and defines a
> +function under `llm-tc/NAME' whose docstring is the value of the spec
> +`:description'.  The tool is then added to
> +`llm-tool-collection--all-tools'."
>
> If this is in the docstring, should we make it into a public variable (as
> in, llm-tool-collection-all-tools)? I originally intended for this to be
> an internal variable only (for llm-tool-collection-get-all), but I could
> see how a list of the symbols could be useful too.
> ------------------------------
>
> In llm-tool-collection.el
> <https://github.com/skissue/llm-tool-collection/pull/2#discussion_r2008773330>
> :
>
> >
> -(defmacro llm-tool-collection-deftool (name specs arguments &rest body)
> +(defvar llm-tool-collection-after-tool-define-hook nil
>
> The Emacs Lisp manual recommends
> <https://www.gnu.org/software/emacs/manual/html_node/emacs/Hooks.html>
> that we use -functions instead of -hook for "abnormal" hooks (those that
> take arguments).
> ------------------------------
>
> In llm-tool-collection.el
> <https://github.com/skissue/llm-tool-collection/pull/2#discussion_r2008773905>
> :
>
> > +    (setq arg-syms (reverse arg-syms)
> +          arg-specs (reverse arg-specs))
> +    (let* ((sym (llm-tool-collection--name-to-symbol name))
> +           (name-spec (unless (plist-get specs :name)
> +                        `(:name ,(llm-tool-collection--make-llm-name name)))))
> +      `(progn
> +         (defconst ,sym
> +           ***@***.***
> +             ***@***.***
> +             :args ,arg-specs
> +             :function ,sym))
> +         (defun ,sym ,arg-syms
> +           ,(concat (plist-get specs :description) "\n\n"
> +                    "Definition generated by `llm-tool-collection'.")
> +           ***@***.***)
> +         (add-to-list 'llm-tool-collection--all-tools ',sym)
>
> The docstring for add-to-list states:
>
> This is meant to be used for adding elements to configuration
> variables, such as adding a directory to a path variable
> like load-path, but please do not abuse it to construct
> arbitrary lists in Elisp code, where using push or cl-pushnew
> will get you more efficient code.
>
> Thus, would it be better to use cl-pushnew instead? I know you explicitly
> wanted to avoid the dependency on cl-lib, but I don't think it's a big
> deal, given that it would be a compile-time dependency. However, I'm
> willing to defer to popular convention here—if add-to-list is commonly
> used in packages, then we can follow others' example. I suppose this is
> also a minor, small use, so perhaps I'm overreacting :).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/skissue/llm-tool-collection/pull/2#pullrequestreview-2708031939>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APC5ZXOHZ5OONXQV7TWTH5L2VVSAFAVCNFSM6AAAAABZR4LCBWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDOMBYGAZTCOJTHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

skissue commented:
> for the commit messages, I’d have a slight preference for adopting the GNU ChangeLog style

To be completely honest, I didn't know this was a thing! I've been using Conventional Commits purely out of habit and experience—I'll definitely take a look at the GNU style when I have a moment.

ultronozm commented:
> > for the commit messages, I’d have a slight preference for adopting the GNU ChangeLog style
> 
> To be completely honest, I didn't know this was a thing! I've been using Conventional Commits purely out of habit and experience—I'll definitely take a look at the GNU style when I have a moment.

Very good.  These are described in the CONTRIBUTE file in my Emacs installation, and maybe also yours.  I'll emphasize that I'm just "following the herd" here and have no strong opinions on commit conventions (besides wanting to be lazy and fit in).

ultronozm commented:
I made one further tweak just now, converting category strings to symbols.  The rationale is that they're enum-like (belonging to some finite collection of values rather than being general strings), so it's more idiomatic (and more efficient, I suppose) to treat them as symbols rather than strings.  I'll be happy to revise if you feel otherwise

ultronozm commented:
OK, one more little tweak (and sorry for the spam).  Since `:description` is an essentially mandatory field for functions and args, it seemed natural to drop the `:description` key and instead place its value immediately after the symbol name.  This makes it look more like an actual docstring...

skissue commented:
> I made one further tweak just now, converting category strings to symbols. The rationale is that they're enum-like (belonging to some finite collection of values rather than being general strings), so it's more idiomatic (and more efficient, I suppose) to treat them as symbols rather than strings. I'll be happy to revise if you feel otherwise

I agree that symbols would be more appropriate. However, the reason that it is a string is because `:category` is actually a recognized keyword by gptel, and gptel takes it as a string. Thus, our hand is forced here. I am planning to add support for "tags" as an organization scheme as well; since that will be arbitrary, I will likely make those symbols.

> OK, one more little tweak (and sorry for the spam). Since `:description` is an essentially mandatory field for functions and args, it seemed natural to drop the `:description` key and instead place its value immediately after the symbol name. This makes it look more like an actual docstring...

No worries—I agree, these are good changes!

ultronozm commented:
> I agree that symbols would be more appropriate. However, the reason that it is a string is because `:category` is actually a recognized keyword by gptel, and gptel takes it as a string. Thus, our hand is forced here. I am planning to add support for "tags" as an organization scheme as well; since that will be arbitrary, I will likely make those symbols.

OK, I see, thanks for clarifying.  I've switched it back to strings.

I've checked that the changes needed in `gptel` to work with symbols (or both) are quite straightforward, so if you'd like, I could submit a PR there and see how karthink feels about it.  I'd also be happy to leave things as they are, if you prefer.

One point concerning optional arguments might be worth discussing.  With the LLM's, I think any of the arguments may be marked as required or optional?  With Emacs, the required must come before the optional.  I updated the docstring of `llm-tool-collection-deftool` to hopefully clarify that point a bit.

ultronozm commented:
One other thought that came to mind is whether it would make sense to include commands in `llm-tool-collection` that make it easier to work on tools by having frontends update as soon as the deftool form is evaluated.

For instance, the following code should do the trick with `gptel`:

```elisp
(defun llm-tool-collection-register-with-gptel (tool-spec)
  "Register a tool defined by TOOL-SPEC with gptel.
TOOL-SPEC is a plist that can be passed to `gptel-make-tool'."
  (when (featurep 'gptel)
    (declare-function gptel-make-tool "gptel")
    (declare-function gptel-tool-name "gptel")
    (defvar gptel-tools)
    (let ((tool (apply #'gptel-make-tool tool-spec)))
      (setq gptel-tools
            (cons tool (seq-remove
                        (lambda (existing)
                          (string= (gptel-tool-name existing)
                                   (gptel-tool-name tool)))
                        gptel-tools))))))

(add-hook 'llm-tool-collection-post-define-functions
          #'llm-tool-collection-register-with-gptel)
```

It's not clear to me whether such code belongs in the README, in some docstring, as part of `llm-tool-collection`, as part of `gptel`, or wherever else, but figured I'd share it for the sake of discussion.

For my personal package [ai-org-chat.el](https://github.com/ultronozm/ai-org-chat.el), the corresponding code is
```elisp
(add-hook 'llm-tool-collection-post-define-functions #'ai-org-chat-register-tool-spec)
```

karthink commented:
> > I agree that symbols would be more appropriate. However, the reason that it is a string is because `:category` is actually a recognized keyword by gptel, and gptel takes it as a string. Thus, our hand is forced here. I am planning to add support for "tags" as an organization scheme as well; since that will be arbitrary, I will likely make those symbols.
> 
> OK, I see, thanks for clarifying. I've switched it back to strings.
> 
> I've checked that the changes needed in `gptel` to work with symbols (or both) are quite straightforward, so if you'd like, I could submit a PR there and see how karthink feels about it. I'd also be happy to leave things as they are, if you prefer.

I made the category a string to give the user freedom to name tool groups however they wanted.  While I expect that a few standard categories will arise over time, users can also have category names like
- `"retrieval (local, text)"`
- `"retrieval (Roam db + Pocket)"`
- `"retrieval (web sources)"`
- `"slack and irc"`
- `"filesystem, read-only"`
- `"eval (watch out!)"`
Users have different ways to tag, and not all of these can be represented as symbols.  So I wasn't thinking of it as an enum.

I agree that tools contributed here should use be categorized from some semi-standard set (like an enum).

karthink commented:
> ```emacs-lisp
> ```emacs-lisp
> 
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> It's not clear to me whether such code belongs in the README, in some docstring, as part of llm-tool-collection, as part of gptel, or wherever else, but figured I'd share it for the sake of discussion.
> ```

It can be part of gptel if required.  Unlike the more commercial offerings, we're developing this split across several projects (gptel, llm, llm-tool-collection, mcp) so I'm in favor of providing glue code in all these packages to make it easy for the user to integrate them.

karthink commented:
> OK, one more little tweak (and sorry for the spam). Since :description is an essentially mandatory field for functions and args, it seemed natural to drop the :description key and instead place its value immediately after the symbol name. This makes it look more like an actual docstring...

See also https://github.com/karthink/gptel/pull/685

ultronozm commented:
> As an aside, for elisp tools it usually makes sense to have a function docstring that's different from the description intended for the LLM. The latter focuses on different things. For example, it can contain instructions for the LLM on which tool (or set of tools) should be called next depending on the result of this one. This helps guide chained tool use with LLMs, as in [this video](https://www.youtube.com/watch?v=2VoOoS4cEV0) and the previous one in this series.

Thanks for this and your other comments.  That's a good point about the different purposes of the docstrings, so I guess it'd make sense to add a further argument specifying the docstring.  It's not clear to me what would be the best ergonomics for this, so I'll leave it for now.

skissue commented:
> One point concerning optional arguments might be worth discussing. With the LLM's, I think any of the arguments may be marked as required or optional? With Emacs, the required must come before the optional. I updated the docstring of `llm-tool-collection-deftool` to hopefully clarify that point a bit.

That's a good point. I don't think it should be an issue—is there ever a scenario where required and optional arguments need to be intermixed?

> One other thought that came to mind is whether it would make sense to include commands in `llm-tool-collection` that make it easier to work on tools by having frontends update as soon as the deftool form is evaluated.
> 
> It's not clear to me whether such code belongs in the README, in some docstring, as part of `llm-tool-collection`, as part of `gptel`, or wherever else, but figured I'd share it for the sake of discussion.

Personally, I don't think I'll be adding this to the codebase itself. To me, users that are only consuming—not iterating on—the tools here should never have a reason to be rapidly evaluating the same tool many times. They can instead add all tools of interest via functions such as `llm-tool-collection-get-all`. Additionally, I'd like this collection to stay reasonably client-agnostic (though the `:category` tag shows that sometimes that does have to be traded for pragmatism). However, this is definitely a good thing to have for development, and I will certainly add this to the README when I get around to writing that (hopefully after merging this PR!).

> > As an aside, for elisp tools it usually makes sense to have a function docstring that's different from the description intended for the LLM. The latter focuses on different things. For example, it can contain instructions for the LLM on which tool (or set of tools) should be called next depending on the result of this one. This helps guide chained tool use with LLMs, as in [this video](https://www.youtube.com/watch?v=2VoOoS4cEV0) and the previous one in this series.
> 
> Thanks for this and your other comments. That's a good point about the different purposes of the docstrings, so I guess it'd make sense to add a further argument specifying the docstring. It's not clear to me what would be the best ergonomics for this, so I'll leave it for now.

Perhaps this is naive or reductive of me, but I am of the (tentative) opinion that, in this specific case, we can simply forgo the Elisp docstring entirely. To me, the most important benefit of the conventions of an Elisp docstring is ensuring that the interface is well defined: arguments, return value, and context. I feel that LLM-friendly descriptions almost always end up being a superset of the information one would have gleaned from an Elisp docstring, since the former also requires documenting context, arguments, and the return value. However, I'm open to examples of this assumption being false.

skissue commented on llm-tool-collection.el:
```diff
@@ -31,93 +31,205 @@
+  "Functions called after defining a new LLM tool.
+Each function is called with one argument, the tool's plist definition.")
+
+;;;###autoload
```
Is there a reason for this macro to be autoloaded? I'm not sure if there is application outside of this package—I was under the impression that this was a macro solely to simplify making definitions for us (and contributors).

ultronozm replied:
This seemed necessary at some point to get font-lock and
indentation for the deftool macros working when editing
llm-tool-collection.el before loading the package, but I haven't been able
to reproduce that issue, so I'll be happy to remove the autoload (but will
first wait for your feedback on the other commits).

skissue replied:
Other commits look good to me; there's a little wording I'd like to tweak, but it's not an issue at all and I'm happy to do it post-merge. Should be ready to merge after this :tada:!

ultronozm replied:
Thanks, updated, should be all set

ultronozm commented:
Sounds good on keeping things package agnostic, and forgoing Elisp
docstrings for now.

On the autoload: this seemed necessary at some point to get font-lock and
indentation for the deftool macros working when editing
llm-tool-collection.el before loading the package, but I haven't been able
to reproduce that issue, so I'll be happy to remove the autoload (but will
first wait for your feedback on the other commits).

On Mon, Mar 24, 2025 at 5:19 PM Ad ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In llm-tool-collection.el
> <https://github.com/skissue/llm-tool-collection/pull/2#discussion_r2010505334>
> :
>
> >
> -(defmacro llm-tool-collection-deftool (name specs arguments &rest body)
> +(defvar llm-tool-collection-post-define-functions nil
> +  "Functions called after defining a new LLM tool.
> +Each function is called with one argument, the tool's plist definition.")
> +
> +;;;###autoload
>
> Is there a reason for this macro to be autoloaded? I'm not sure if there
> is application outside of this package—I was under the impression that this
> was a macro solely to simplify making definitions for us (and contributors).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/skissue/llm-tool-collection/pull/2#pullrequestreview-2710963824>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APC5ZXKNFR6SWPNR4OJSOGT2WAWABAVCNFSM6AAAAABZR4LCBWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDOMJQHE3DGOBSGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

skissue commented:
Thank you so much!
